[CIP 0007] WIP Downtime Slasher (#9)
celo-org/CIPs: CIPs/0007.md

# CIP [0007]: Affordable Gas Downtime Slasher

- Date: 2020-05-05
- Author: @gastonponti (with a lot of feedback and ideas from @m-chrzan)
- Status: WIP

## Overview

> This proposal contains a new DowntimeSlasher contract.

Author: Totally agree

> The actual DowntimeSlasher tries to verify that the validator signed at least a block in all the window that is required to be validated which it leads to running out of gas for big windows.

Reviewer: We would like any readers unfamiliar with the issue that prompted this proposal, and who may be external to cLabs, to have the context to understand and provide comments on the proposal.
```suggestion
Currently, slashable downtime is defined as a window of blocks, greater than a predefined length, in which no block's parent aggregated seal includes a signature from the validator in question. It's currently implementation checks and executes slashing in a single transaction which requires an amount of gas linear in the required window size.

Under its current implementation, the gas required to evaluate a downtime event may far exceed the block gas limit. In the Release Candidate 1 network, the downtime slashing window is configured to be 5760 blocks, which would require roughly 13.5 million gas to evaluate, while the block gas limit is set to 10 million.
```

> The main goal of this new contract is to generate a tool that would allow any user to slash a validator for every possible block windows without gas limitations. 
> 
> For the actual window of 1/2 epoch, the DowntimeSlasher won't finish with a "run out gas" exception. Without this new tool for slashing, there is no a possible way to penalize a validator for downtime.
> 

Reviewer: ```suggestion
```
See above comments.

## Goals

Make a new contract that:
- Allows the user to slash a validator if it was down for a specific consecutive block window

> - The implementation works for every possible block window (required in the initialization)

Reviewer: What does this mean? Could you rephrase this?

## Proposed Solution

Create a new contract funtion that divides the downtime verification in valid slots.

The user that wants to slash a particular validator will have to call a function from the contract `S_n` times (`S_n` could be asked to the contract itself) to verify each slot.

Then the user must generate another contract call to verify that the validator to be slashed was down for every consecutive slot.

### Process

1) The user calls the contract to know how many slots are required (`S_n`) and how many blocks must have every slot (`B_n`)

2) The user requires to validate `S_n` slots
    - The user calls the contract for the slot validation `[StartBlockNumber + (i * B_n), StartBlockNumber + ((1 + i) * B_n) )` with `0 <= i < S_n`

3) The user calls the slash method with the `StartBlockNumber` (internally will verify that every slot validation was performed and succeded)

> ### Variation of the proposed solution

Reviewer 2: Which of these are you proposing?

Author: Well, I'm not totally decided, and I want to discuss it a little.

For my point of view, the "user has to verify everything" it's the approach that would be easier for the community to accept because it's not changing anything (same as the old one).
That's why I was thinking to use that. But I would like some opinions on that.

(And you are totally right, I should write which option I'm proposing clearer)

Reviewer 3: I like the 1st proposal.  It's much more intuitive (which will incidentally make implementation much more easier) and I would think that the added functionality of the 2nd proposal will hardly be utilized.  E.g. it will almost always be one one user that generates all of the validation slots.  I suppose that we could also upgrade the smart contract later on down the road if there is alot of community demand for the 2nd proposal.

- Every user generates the validation slots for itself. The first user to gather all the validations and trigger the slashing, collects the reward

- The slots validated works for every user in the community. When a slashing is performed, every user that generated a validated slot that is used to demostrate the downtime, receives a portion of the reward 

## Alternative Solutions

### Validator's Uptime in the epoch

> The Celo blockchain for reward purposes, calculates an uptime for each validator. The validator’s uptime for the epoch is the proportion of blocks in the epoch for which it is ‘up’: U = counter / (epoch_size - 11).

Reviewer: The uptime number also uses the 12-block sliding window, so this description is not quite accurate.

> One possible approach, it's to slash the validators if they are not covering a certain umbral.

Reviewer: ```suggestion
One possible approach, is to slash the validators if they do not meet a certain threshold for uptime in an epoch.
```

#### Pros

The uptime calculation it's part of the Celo protocol. This not only allow us to save gas from calculating a possible slashing, but also could let the same blockchain to perform the slashes. This would make a safer net, instead of requiring a so rigurous control of the users.

In the actual scenario, if no user is controling, no one get slashed.

> #### Cons

Reviewer 3: Seems like one way to address most (if not all the cons) is that every validator will check for uptime at *every* block, and not just at the end of the epoch.

The validator's uptime is calculated at the end of the epoch, that has two potential problems:

- The slashing for an specific validator will be performed in the next epoch. That allows a validator that is not performing well in the epoch X, to be reelected for the epoch X+1, because the tool used for that (the uptime) was added after the election was completed.

- The uptime is a relation that has meaning tied to an specific epoch as whole. It won't give you a detail information of what happened.

> For example, if the community sets the lower umbral as 50% this two scenarios have the same uptime:

Reviewer: English work is "threshold" :)
```suggestion
For example, if the community sets the lower threshold as 50% this two scenarios have the same uptime:
```

 * A Validator is down for 1/2 epoch consecutive, and then signs every block for the next 1/2 epoch
 * A Validator signs 1 every 2 packages

Another problem is the calculation between epoch for reelected validators.
Using the same umbral of the other example, if the validator has in the epoch X = 75% uptime, and in the epoch X+1 = 70% uptime, this two numbers could hide a more that a 1/2 epoch consecutive blocks downtime. This will happen if the 25% of the downtime in the X is at the ending of the epoch, and the other 30% of the downtime for the epoch X+1 is at the beginning. 

### Challenge the validator

This solutions consist in solving a challenge between a user and the validator. 

The user could claim that the validator has been down for an specific period of blocks. In which case, the validator has an specific amount of time to bring a proove where the blockchain could verify that the validator was up. For example 1 or N blocks that where signed by the validator.

#### Pros

> From the blockchain perspective, a simple and easy way for validating the uptime.

Reviewer: Implementing a challenge-response scheme would be far from simple in practice. I don't think we have a fleshed out enough idea of what this would look like to really include it as a proposal in this CIP.

# CIP [0007]: Affordable Gas Downtime Slasher

- Date: 2020-05-05
- Author: @gastonponti (with a lot of feedback and ideas from @m-chrzan)

> - Status: Implemented

Reviewer 4: This needs to be changed to Draft status

Reviewer: Just to make sure we are on the same page, this is implemented and being deployed with contract release 1. Is the status flipped once the contract is finally deployed?

Author: Exactly

## Overview

This proposal includes improvements to Celo's downtime slashing, such that it can be evaluated within the block gas limit.

In order to promote a stable network, the Celo protocol includes slashing for downtime such that if a validator fails to participate in consensus for a given amount of time, they may be slashed. More information about slashing in the Celo protocol can be found on [docs.celo.org](https://docs.celo.org/celo-codebase/protocol/proof-of-stake/penalties)

> Currently, slashable downtime is defined as a window of blocks, greater than a predefined length, in which no block's parent aggregated seal includes a signature from the validator in question. It's currently implementation checks and executes slashing in a single transaction which requires an amount of gas linear in the required window size.

Reviewer: The definition did not change. Removing the work "currently" would help clarify that.
```suggestion
Slashable downtime is defined as a window of blocks, greater than a predefined length, in which no block's parent aggregated seal includes a signature from the validator in question. It's currently implementation checks and executes slashing in a single transaction which requires an amount of gas linear in the required window size.
```

Under its current implementation, the gas required to evaluate a downtime event may far exceed the block gas limit. In the Release Candidate 1 network, the downtime slashing window is configured to be 5760 blocks, which would require roughly 13.5 million gas to evaluate, while the block gas limit is set to 10 million.

## Goal

- Allow the user to slash a validator if it was down for a specific consecutive block window without running out of gas.

## Proposed Solution

> Create a new contract funtion that divides the downtime verification in validation intervals.

Reviewer 4: function* typo

The user that wants to slash a particular validator will have to call a function from the contract that will generate a proof with the validators that signed at least one block for the specific interval set by the caller.

The user will require to repeat this first process the amount of times necessary to cover with overlapping or adjacent intervals, at least the `slashableDowntime` amount of blocks set in the contract.

Then the user must generate another contract call listing all the intervals required to verify that the validator to be slashed was down for those intervals, and that also covers the `slashableDowntime` window.

### Note

> This process allows the user to validate a interval that will contain the data of every validator. This means that if for example, three validators where down in the same period, the same interval could be used to slash the three of them.

Reviewer 4: an* interval

> ### Note
> 
> This process allows the user to validate a interval that will contain the data of every validator. This means that if for example, three validators where down in the same period, the same interval could be used to slash the three of them.
> The gas required to generate the `I_n` intervals will be spent only in the first validation.

Reviewer: ```suggestion
> Note: This process allows the user to validate a interval that will contain the data of every validator. This means that if for example, three validators where down in the same period, the same interval could be used to slash the three of them.
> The gas required to generate the `I_n` intervals will be spent only in the first validation.
```

### Some numbers

Tested with intervals of 5000 blocks (each one required less than 10 million gas), using 1000 intervals for the actual slash (required ~6.5 million gas) which covers 5M blocks.
Even if is not necessary right now, this solution will cover possible governable incrementations of the `slashableDowntime` window.

### Variation of the proposed solution

- The slots validated could work for every user in the community. When a slashing is performed, every user that generated a validated slot that is used to demostrate the downtime, receives a portion of the reward. But to address this, changes are required in other contracts to calculate and distribute the rewards.

> ## Alternative Solution

Reviewer 4: I would remove this entire section. Alternative solutions should be proposed under their own respective CIPs

Reviewer 4: I now see that Alternative Solutions was specified in the current template, so no worries, please keep this section :)
